### ISSUE TITLE
Enhance support.reap_children()

### ISSUE THREAD

**Issue by vstinner** [python/cpython#75343]
BPO | [31160](https://bugs.python.org/issue31160)
--- | :---
Nosy | @vstinner, @aixtools
PRs | <li>python/cpython#3036</li><li>python/cpython#3040</li><li>python/cpython#3043</li><li>python/cpython#3044</li><li>python/cpython#3045</li><li>python/cpython#3046</li><li>python/cpython#3056</li><li>python/cpython#3060</li><li>python/cpython#3063</li><li>python/cpython#19263</li><li>python/cpython#19308</li><li>python/cpython#19312</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-08-11.00:12:50.391>
created_at = <Date 2017-08-09.12:50:03.350>
labels = []
title = 'Enhance support.reap_children()'
updated_at = <Date 2020-04-02.20:08:02.894>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2020-04-02.20:08:02.894>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-08-11.00:12:50.391>
closer = 'vstinner'
components = []
creation = <Date 2017-08-09.12:50:03.350>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 31160
keywords = []
message_count = 23.0
messages = ['299996', '299997', '299998', '300000', '300020', '300021', '300023', '300024', '300072', '300073', '300086', '300088', '300093', '300147', '365430', '365474', '365475', '365476', '365477', '365478', '365479', '365512', '365517']
nosy_count = 2.0
nosy_names = ['vstinner', 'Michael.Felt']
pr_nums = ['3036', '3040', '3043', '3044', '3045', '3046', '3056', '3060', '3063', '19263', '19308', '19312']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue31160'
versions = []
```

</p></details>

**Comment by vstinner** [comment]
Attached PR enhances the support.reap_children() function:

* reap_children() now sets environment_altered to True to detect bugs using python3 -m test --fail-env-changed
* Replace bare "except:" with "except OSError:" in reap_children()
* Write an unit test for reap_children() using a timeout of 60 seconds

**Comment by vstinner** [comment]
The GCC job of Travis CI failed with ENV_CHANGED:

Test wait() behavior when waitpid returns WIFSTOPPED; bpo-29335. ...
Warning -- reap_children() reaped child process 19839
ok

I tested and... WOW! When run in a loop, this test leaks 100 MB per second. It creates a lot of processes.

**Comment by vstinner** [comment]
> The GCC job of Travis CI failed with ENV_CHANGED:

Ooops, in fact it was a macOS job:
https://travis-ci.org/python/cpython/jobs/262606830


The GCC job failed with much more errors:
https://travis-ci.org/python/cpython/jobs/262606831

\---
0:00:02 load avg: 121.34 [  7/403] test_unittest
Warning -- reap_children() reaped child process 11088
Warning -- reap_children() reaped child process 11089
Warning -- reap_children() reaped child process 11090
Warning -- reap_children() reaped child process 11091

0:14:22 load avg: 136.67 [282/403/1] test_select
Warning -- reap_children() reaped child process 14686

0:16:00 load avg: 110.41 [297/403/1] test_socketserver
Warning -- reap_children() reaped child process 15483
Warning -- reap_children() reaped child process 15492
Warning -- reap_children() reaped child process 15499
Warning -- reap_children() reaped child process 15508

0:18:36 load avg: 105.94 [333/403/1] test_thread
Warning -- reap_children() reaped child process 20670
\---

For test_socketserver, see bpo-31151.

**Comment by vstinner** [comment]
A reap_children() warning was fixed in test_thread: bpo-31150. It seems like the commit 88eee44a91c991dbbdf284fa220e2928b5de105c was not enough to fix all warnings.

**Comment by vstinner** [comment]
New changeset 4baca1b0f7325032598cd38e7ceffc79b616d255 by Victor Stinner in branch 'master':
bpo-31160: Fix test_builtin for zombie process (bpo-3043)
https://github.com/python/cpython/commit/4baca1b0f7325032598cd38e7ceffc79b616d255

**Comment by vstinner** [comment]
New changeset e3510d74aacc477c30f42f2b941d69689bbc478e by Victor Stinner in branch 'master':
bpo-31160: regrtest now reaps child processes (bpo-3044)
https://github.com/python/cpython/commit/e3510d74aacc477c30f42f2b941d69689bbc478e

**Comment by vstinner** [comment]
New changeset da5e9301877346942fa4279612750d6a09e05153 by Victor Stinner in branch 'master':
bpo-31160: Fix test_random for zombie process (bpo-3045)
https://github.com/python/cpython/commit/da5e9301877346942fa4279612750d6a09e05153

**Comment by vstinner** [comment]
> New changeset 4baca1b0f7325032598cd38e7ceffc79b616d255 by Victor Stinner in branch 'master':
> bpo-31160: Fix test_builtin for zombie process (bpo-3043)

This change introduced a regression:

http://buildbot.python.org/all/builders/AMD64%20Debian%20root%203.x/builds/1159/steps/test/logs/stdio

\======================================================================
FAIL: test_input_no_stdout_fileno (test.test_builtin.PtyTests)
\----------------------------------------------------------------------
```pytb
Traceback (most recent call last):
  File "/root/buildarea/3.x.angelico-debian-amd64/build/Lib/test/test_builtin.py", line 1624, in test_input_no_stdout_fileno
    lines = self.run_child(child, b"quux\r")
  File "/root/buildarea/3.x.angelico-debian-amd64/build/Lib/test/test_builtin.py", line 1573, in run_child
    self.assertEqual(status, 0)
AssertionError: 1 != 0
```

**Comment by vstinner** [comment]
bpo-31173 fixed a leaked child process in test_subprocess.

**Comment by vstinner** [comment]
New changeset 6c8c2943d996b59a48d331f61f22cbe72933910e by Victor Stinner in branch 'master':
bpo-31160: test_tempfile: Fix reap_children() warning (bpo-3056)
https://github.com/python/cpython/commit/6c8c2943d996b59a48d331f61f22cbe72933910e

**Comment by vstinner** [comment]
bpo-31151 fixed test_socketserver.

**Comment by vstinner** [comment]
New changeset 719a15b32587de6c4add4385ee3f84a20711340f by Victor Stinner in branch '3.6':
[3.6] bpo-31160: Backport reap_children() fixes from master to 3.6 (bpo-3060)
https://github.com/python/cpython/commit/719a15b32587de6c4add4385ee3f84a20711340f

**Comment by vstinner** [comment]
New changeset 1247e2cda514d7a73187e0b53ec8c35d87a34a84 by Victor Stinner in branch '2.7':
[2.7] bpo-31160: Backport reap_children fixes from master to 2.7 (bpo-3063)
https://github.com/python/cpython/commit/1247e2cda514d7a73187e0b53ec8c35d87a34a84

**Comment by vstinner** [comment]
I pushed the most important change: reap_children() now makes tests fail with ENV_CHANGED on warning, so I close the issue.
